### ISSUE TITLE
Losing PV while PVC still exists

### ISSUE THREAD

**Issue by Calpicow** [kubernetes/kubernetes#48081]
I'm losing PVs while their attached PVCs still exist. Happens about once a month since upgrading to v1.5.x. Using kube v1.5.3 in AWS GovCloud.

```
W0522 01:00:23.537501 1 listers.go:69] can not retrieve list of objects using index : Index with name namespace does not exist
I0522 01:00:23.691390 1 pv_controller.go:441] volume "pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3" is released and reclaim policy "Delete" will be executed
I0522 01:00:23.897945 1 panics.go:76] PUT /api/v1/persistentvolumes/pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3/status: (113.465551ms) 200 [[hyperkube/v1.5.3 (linux/amd64) kubernetes/029c3a4/persistent-volume-binder] 127.0.0.1:46558]
I0522 01:00:23.898777 1 pv_controller.go:672] volume "pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3" entered phase "Released"
I0522 01:00:24.097903 1 panics.go:76] GET /api/v1/persistentvolumes/pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3: (63.425359ms) 200 [[hyperkube/v1.5.3 (linux/amd64) kubernetes/029c3a4/persistent-volume-binder] 127.0.0.1:46580]
I0522 01:00:24.098147 1 pv_controller.go:1155] isVolumeReleased[pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3]: volume is released
I0522 01:00:24.558468 1 aws_util.go:59] Successfully deleted EBS Disk volume aws://us-gov-west-1a/vol-07b171169f4fbdeb6
I0522 01:00:24.558484 1 pv_controller.go:1191] volume "pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3" deleted
I0522 01:00:24.877886 1 panics.go:76] DELETE /api/v1/persistentvolumes/pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3: (193.380685ms) 200 [[hyperkube/v1.5.3 (linux/amd64) kubernetes/029c3a4/persistent-volume-binder] 127.0.0.1:46628]
I0522 01:00:27.037645 1 panics.go:76] GET /api/v1/persistentvolumes/pvc-f7e47a97-3a52-11e7-b904-06d97c0700c3: (53.083937ms) 404 [[hyperkube/v1.5.3 (linux/amd64) kubernetes/029c3a4/persistent-volume-binder] 127.0.0.1:46734]
I0522 01:00:30.350036 1 panics.go:76] PUT /api/v1/namespaces/test/persistentvolumeclaims/testpvc/status: (133.804718ms) 200 [[hyperkube/v1.5.3 (linux/amd64) kubernetes/029c3a4/persistent-volume-binder] 127.0.0.1:46882]
I0522 01:00:30.350373 1 pv_controller.go:608] claim "test/testpvc" entered phase "Lost"
```

/kind bug
/sig storage

**Comment by msau42** [comment]
Do you still see this issue?  Can you also post logs with the corresponding PVCs?

**Comment by Calpicow** [comment]
Haven't seen it since. I've also upgraded to v1.6, so will update if it happens again.

**Comment by msau42** [comment]
Please reopen if you see the issue.

/assign

**Comment by fankangbest** [comment]
I met this issue before when i reboot the whole cluster. When this issue happens, the code in /pkg/controller/volume/persistentvolume/pv_controller_base.go initializeCaches() get an non-nil err but the code just ignores this error。
        claimListObj, err := claimSource.List(api.ListOptions{})
        if err != nil {
                glog.Errorf("PersistentVolumeController can't initialize caches: %v", err)
                return
        }
Using k8s v1.5.2。

**Comment by msau42** [comment]
Can you open a new issue and post logs from the API server and PV controller?

**Comment by fankangbest** [comment]
@msau42 
This issue happened once when we reboot the whole cluster (with three masters for high availability) about forty times. I will try to recurrent this issue with a easier way. Our temporary solution is to ensure there is no err happens when exucuting initializeCaches(), which is located in /pkg/controller/volume/persistentvolume/pv_controller_base.go :
The original code is:
``` Go
// original code
// Run starts all of this controller's control loops
func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
	glog.V(4).Infof("starting PersistentVolumeController")
	ctrl.initializeCaches(ctrl.volumeSource, ctrl.claimSource)
	go ctrl.volumeController.Run(stopCh)
	go ctrl.claimController.Run(stopCh)
	go ctrl.classReflector.RunUntil(stopCh)
}
```
Modified code looks like:
``` Go
// modified pseudo-code
// Run starts all of this controller's control loops
func (ctrl *PersistentVolumeController) Run(stopCh <-chan struct{}) {
	glog.V(4).Infof("starting PersistentVolumeController")
        for {
	      err := ctrl.initializeCaches(ctrl.volumeSource, ctrl.claimSource)
              if err == nil {
                    break
              }
              sleep for a while
        }
	go ctrl.volumeController.Run(stopCh)
	go ctrl.claimController.Run(stopCh)
	go ctrl.classReflector.RunUntil(stopCh)
}
```
Error in initializeCaches() is returned.

If error happens in initializeCaches() when getting pvc list, the code will delete pv.

**Comment by msau42** [comment]
Hi @fankangbest this sounds like a potential error condition that could be unit tested.  Would you like to open up a PR and we can discuss the best fix?

**Comment by msau42** [comment]
cc @jsafrane

**Comment by fankangbest** [comment]
@msau42 
It is wrong for "initializeCaches() failed can cause to delete pv". initializeCaches() is a guarantee for store to cache objects. This issue may be triggered by something wrong with claims store and initializeCaches() also fails. 
I will open up a new issue when i have a further study about this issue, or some one who is familiar with related code can also check out what condition can cause this issue.

**Comment by jsafrane** [comment]
@fankangbest do you have the error that caused initializeCaches() to fail? It must have been logged to controller-manager logs.

**Comment by fankangbest** [comment]
[kube-controller-manager.zip](https://github.com/kubernetes/kubernetes/files/1259885/kube-controller-manager.zip)

**Comment by jsafrane** [comment]
Thanks, I can see:
```
E0613 16:01:54.143568     889 pv_controller_base.go:193] PersistentVolumeController can't initialize caches: the server cannot complete the requested operation at this time, try again later (get persistentvolumeclaims)
```

Question is, why PVs work and the controller can list them and get events and PVCs don't. The first event from a PVC arrived ~26 minutes after start!

Anyway, this bug should be fixed in 1.6.x by https://github.com/jsafrane/kubernetes/commit/e63fcf708d104bf6c69033ca0e2321bd826c2b40
